2.0 training data format docs (#6357)
RasaHQ/rasa: docs/docs/user-guide/training-data-format.mdx

same name when the model is trained. Here is an example of a story that ends
with a checkpoint, and one that starts with the same checkpoint:

```yaml
- story: story_with_a_checkpoint_1
  steps:
  - intent: greet
  - action: utter_greet
  - checkpoint: greet_checkpoint

- story: story_with_a_checkpoint_2
  steps:
  - checkpoint: greet_checkpoint
  - intent: book_flight
  - action: action_book_flight
```

Checkpoints at the beginning of stories can also be conditional on slots being set, for example:

```yaml
- story: story_with_a_conditional_checkpoint
  steps:
  - checkpoint: greet_checkpoint
    # This checkpoint should only apply if slot is set to the specified value
    slots: 
    - context_scenario: holiday
    - holiday_name: thanksgiving
  - intent: greet
  - action: utter_greet_thanksgiving
```

Checkpoints can help simplify your training data and reduce redundancy in it,
but **do not overuse them**. Using lots of checkpoints can quickly make your
stories hard to understand. It makes sense to use them if a sequence of steps
is repeated often in different stories, but stories without checkpoints
are easier to read and write. 

> #### or

Author: ```suggestion
#### OR statement
```

---
id: training-data-format
sidebar_label: Training Data Format
title: Training Data Format
description: Description of the YAML format for training data
---

## Overview

> ### YAML Format

Reviewer: I'd remove this headline and put the text under `overview`. I think we should either add a paragraph or note on old data formats with a link to their documentation

Author: added note; not sure where links should point

> Rasa Open Source 2.0 uses [YAML](https://yaml.org/spec/1.2/spec.html) as

Reviewer: ```suggestion
Rasa Open Source uses [YAML](https://yaml.org/spec/1.2/spec.html) as
```

I don't think we need to mention that if we add a note about the deprecated formats

a unified and extendable way to manage all training data,
including NLU data, stories and rules. 

With the YAML format, training data can be split over any number of YAML files,

> and every file can contain any kind of data. 

Reviewer: @degiz doing the multimedia response implementation I noticed that this is actually not quite true (we should make it work though). We do the "what is this file type containing" discovery (is it nlu, stories, ...) and the parsing afterwards. if your file contains stories and nlu data, only one of them will be read. are you aware of that?

Reviewer 2: If a file contains `nlu` and `stories` key, it should be accepted by both the NLU and Stories parser.
If that's not the case, I need to fix that 😬 I'll check that.

Reviewer: mhm yes I think both will accept them, but I though only one of them will actually get called 🤔  - I'll close this comment though nothing we need to discuss here.

The [domain](../glossary#domain) uses
the same YAML format as the training data and can also be split across
multiple files or combined with training data in one file. The domain includes
the definitions for [responses](../core/domains#responses) and [forms](../core/forms).

> See the [documentation for domains](../core/domains) to learn more.

Reviewer 3: @m-vdb i'm confused, do these links need to have `.mdx` in them or not? Running them locally with the `make install-docs` from `master` and the links including the anchors seem to render fine 🤔

Reviewer: I think it should be according to the docusaurus docs: https://v2.docusaurus.io/docs/markdown-features/#referencing-other-documents

Reviewer 4: you don't have too, but it looks like it's preferred (the only benefit I see from now is that if we want to navigate the doc from github files, but in our case it wouldn't really work anyway).

### High Level Structure

Each file can contain one or more **keys** with corresponding training
data. One file can contain multiple keys, as long as there is not more
than one of a certain key in a single file. The available keys are:

- `version`
- `stories`
- `nlu`
- `rules`

> - `e2e_tests`

Reviewer 3: let's put these in the order they're explained. I'm wondering if it makes sense to not group stories and rules under `conversation training data`, and rather explain what each is at the beginning of its section? because they're on the same level as `nlu` and there's no reference to "conversation training data" here, maybe it would make sense? not sure though

Reviewer 3: Well lol it looks like we just moved away from that 😅

Author: Just switched this to be one heading today @tmbo  suggestion - https://github.com/RasaHQ/rasa/pull/6357/commits/aba2e2e778c0ebc32bccfc5ddd2e89125dc38f4b

It kinda makes sense to group it because NLU is everything for an NLU model, and then conversational is everything for a dialogue management model.

Author: Why can't I see your follow up comment from the files changed tab  🤔 anyway that's fine then.

- `rules`
- `e2e_tests`

All YAML training data files should specify the `version` key in order
to be parsed correctly. Training data files with no `version` specified will be assumed to

> be in the format of the latest version of Rasa Open Source. Training data files with a version greater than is currently

Reviewer: the version is NOT specifying the rasa open source version, but the training data version

Reviewer: I'd rather phrase it as (also using active voice)

"If you don't specify a `version` key in your training data file, Rasa will assume you are using the latest training data format specification."

Author: Is there info somewhere about training data versions?

Reviewer: Well, so here is how it should work technically:
- we do have different training data specifications, each gets a version. first one is 2.0 - if at some point we need to change the training data schema, e.g. because we want to rename of a training data example, we'd increase the version number to 2.1. this is independent of the rasa open source version though. we should only change the version number if the change is not backwards compatible.
- the user specifies the specification their training data adheres to
- if the installed rasa open source version does not support the specified training data version, it should fail. if the user did not specify any format version, rasa open source should use the latest format it supports

Reviewer: hopefully, we'll not need to make a lot of use of that version number

available for Rasa Open Source will be skipped.

### Example

Here's a short example which keeps all training data in a single file:

```yaml
version: "2.0"

stories:
- story: greet and faq
  steps:
  - intent: greet
  - action: utter_greet
  - intent: faq
  - action: respond_faq

rules:
- rule: Greet user
  steps:
  - intent: greet
  - action: utter_greet

responses:
  faq/language:

>   - text: |

Reviewer: I think we should not use blocks for these but rather use simple strings. e.g.
`- text: "I can only do English at the moment"`

Reviewer: it avoids people confusing this with the nlu syntax and doing odd things like adding a `-` in the beginning

Author: Ah right, I should actually remove the responses example entirely because at the top it says they're part of the domain

  - hey there [Sara](name)

- intent: faq/language
  examples: |
  - What language do you speak?
  - Do you only handle english?

e2e_tests:
- user: |
    hey
  intent: greet
- action: utter_greet
- user: |
    what language do you speak
  intent: faq/language
- action: respond_faq

> ```

Reviewer 2: True, that can cause confusion. I can't think of a better solution atm though.

> ```

Author: If it's a single string, can it just be provided as `- user: what language do you speak` ? Or does that also require the weird escaping?

e2e_tests:
- user: |
    hey
  intent: greet
- action: utter_greet
- user: |
    what language do you speak
  intent: faq/language
- action: respond_faq
```

> ## NLU Training Data

Reviewer: should we rather call this something like "User Message Training Data"?

Reviewer: I'd rather like to avoid the term NLU in most areas as the user likely doesn't know what it means

Reviewer: even if we introduce NLU as a term in this section, we should avoid using these terms in the title of that section

Author: I'm not sure about this. If users are looking at the training data format (which will be not at the top the way it is now), I think they've already seen NLU and know what it is. We also talk about the NLU model, so it makes it clear what model it's feeding.

Author: It's true it should be defined first though

NLU training data consists of example user utterances categorized by
**intent**, i.e. what the user is trying to convey or accomplish with their
message. Training examples can also include **entities**. Entities are structured
pieces of information that can be extracted from a user's message. The other components
of NLU training data are there to help the model identify intent and entity
correctly.

NLU training data is defined under the `nlu` key. Items that can be added under this key are:

- [Training examples](#training-examples) grouped by user intent,
  optionally with annotated [entities](#entities)
- [Synonyms](#synonyms)
- [Regular Expression Features](#regular-expression-features)
- [Lookup Tables](#lookup-tables)

> Here is a simple example of each item type:

Reviewer: should we rather split this up and add the examples separately to the above list? I think that makes it easier to understand and map

```yaml
nlu:
- intent: check_balance
  examples: |
  - What's my [credit](account) balance?
  - What's the balance on my [credit card account]{"entity":"account","value":"credit"}

- synonym: credit
  examples: |
  - credit card account
  - credit account

- regex: account_number
  examples: |
  - \d{10:12}

- lookup: banks
  examples: |
  - JPMC
  - Comerica
  - Bank of America
```

### Training Examples

Training examples are grouped by [intent](../glossary#intent) and listed under the
`examples` key. Usually, you'll list one example per line as follows:

```yaml
nlu:
- intent: greet
  examples: |
  - hey
  - hi
  - whats up
```

> However, its also possible to use an extended format if you want to include

Reviewer: ```suggestion
However, its also possible to use an extended format if you need to include
```

metadata for your examples:

```yaml
nlu:
- intent: greet
  examples: 
  - text: |
      hi
    metadata:
      sentiment: neutral
  - text: |
      hey there!
```

The `metadata` key can contain arbitrary key-value data and the parser will not read its value. 

> Metadata can be useful for storing information relevant to an example, but not used by 

Reviewer: I don't think this is actually true, e.g. the sentiment in the above example would be used in an NLU component (we just don't have one by default).

Author: Ok, I was wondering about that. Is this true then?
```
The `metadata` key can contain arbitrary key-value data. The parser will read it's value, but it will only be used if a component in the pipeline needs it. In the example above, the sentiment of 
the example is recorded as metadata. This could be used by a custom component in the pipeline for sentiment analysis.
```

Reviewer: yes exactly, that makes more sense. metadata really is more an extension point for people that want to use their own or third party components that need additional data annotations

the NLU pipeline. In the example above, the sentiment of 
the example is recorded as metadata.

> ### Entities

Reviewer: should we add a description what entities are here first?

Author: There's this at the top of the nlu section - should it move?
```
Training examples can also include **entities**. Entities are structured
pieces of information that can be extracted from a user's message.
```

Reviewer: Let's keep it there and add some version of that here as well, maybe something along the lines of:
`Entities are structured pieces of information that can be extracted from a user's message. For the entity extraction to work, you need to either specify training data to train an ML model to recognize your entities or you need to define rules to extract entities based on a character pattern.`. not sure if it makes sense to include the ml / rules part, up to you

Where applicable, [entities])(../glossary#entity) can be annotated in training examples using either the
abbreviated syntax:

> ```yaml

Reviewer: I think we should discuss if we should really add both of them or if it would be better to just mention the more flexible one using `{}`

Author: imo, if both are possible still, that should be mentioned. People coming from 1.x might try it anyway and be confused if they see it works but there's no docs for it.

Reviewer: well, it will work but we'll throw a deprecation warning. if we mention both formats, we should at least differentiate them. e.g. adding a note that there is an additional format that is deprecated.

Author: Did not know it'd throw a warning - should've checked. In that case makes sense to remove it.

Reviewer: well, not sure it does already throw that warning. can you either check or create an issue + assign sasha to the issue?

Author: it does not throw a warning, created an issue https://github.com/RasaHQ/rasa/issues/6365

[<entity-text>]("<entity name>")
```

Or the syntax:

```yaml
[<entity-text>]{"entity": "<entity name>"}
```

In a training example, this would look like:

```yaml
nlu:
- intent: check_balance
  examples: |
  - how much do I have on my [savings]("account") account
  - how much money is in my [checking]{"entity": "account"} account

>   - What's the balance on my [credit card account]{"entity":"account","value":"credit"}

Reviewer: I think we should always keep this in mind and use this structure throughout our docs:
explanation first, then an example block.

The above has a structure that is more like
explanation, example with additional syntax, explanation of the additional syntax

This makes jumping through sections harder in my opinion. I'd rather remove the third example and introduce it separately in the paragraph that focuses on it.

Author: 👍 switched it around

```

The third example above includes a [synonym](#synonyms).
You can assign synonyms, roles, or groups to an entity using the
syntax:

```yaml
[<entity-text>]{"entity": "<entity name>", "role": "<role name>", "group": "<group name>", "value": "<entity synonym>"}
```

The keywords `role`, `group`, and `value` are optional in this notation.
The `value` keyword refers to synonyms, which are explained in the
following section. To understand what the labels `role` and `group` are
for, see the section on (entity roles and groups)[nlu/entity-extraction#entities-roles-and-groups].

### Synonyms

Synonyms provide a way to normalize your training data by mapping an
extracted entity to a value other than the literal text extracted.
Synonyms can be defined in the format:

> ```yaml

Reviewer: let's rather use an actual example here

- synonym: <synonym value>
  examples: |
  - <a synonym variation>
  - <another synonym variation>
```

Synonyms can also be definedin-line in your training examples by
specifying the `value` of the entity:

```yaml
nlu:
- intent: check_balance
  examples: |
  - how much do I have on my [credit card account]{"entity": "account", "value": "credit"}
  - how much do I owe on my [credit account]{"entity": "account", "value": "credit"}
```

To use the synonyms defined in your training data, you need to make sure the

> pipeline contains the [`EntitySynonymMapper`](../nlu/components#entitysynonymmapper) component. You

Reviewer: should mention the configuration, as in "pipeline in your configuration file"

```yaml
- synonym: credit
  examples: |
  - credit card account
  - credit account
```

Then, if either of these phrases is extracted as an entity, it will be
mapped to the **value** `credit`.

:::note
Synonym mapping only happens **after** entities have been extracted.
That means that you need to provide examples of the variations 
on a synonym in your training examples so
that Rasa Open Source can learn to pick them up.

> :::

Reviewer 3: How about following the example: 

Synonym mapping only happens **after** entities have been extracted.
That means that your training examples should include the synonym examples (`credit card account` and `credit account`) so that the model will learn to recognize these as entities and replace them with `credit`.

### Regular Expression Features

Regular expressions can be used in two different ways:

1. They can be used to support intent classification and entity

> extraction when using the [RegexFeaturizer](../nlu/components#regexfeaturizer) component in the pipeline.

Reviewer 3: ```suggestion
extraction when using the [`RegexFeaturizer`](../nlu/components#regexfeaturizer) component in the pipeline.
```

2. They can be used to extract entities from a user message using the rule-based
[`RegexEntityExtractor`](../nlu/components#regexentityextractor) component in the pipeline.

For example, if your entity has a deterministic structure (like a

> zipcode or an email address), you can use a regular expression to ease

Reviewer 3: ```suggestion
zipcode or an email address), you can use a regular expression to improve
```

detection of that entity (using the `RegexFeaturizer`) or to
extract the entities from the user message directly (using the
`RegexEntityExtractor`). For a bank account number that is 10-12 digits long,
it might look like this:

```yaml
- regex: account_number
  examples: |
  - \d{10:12}
```

If you are using regular expressions to directly extract entities using
the `RegexEntityExtractor`, the name of the regular expression should
match the name of the entity you want to extract.

If you are using the regular expressions for the `RegexFeaturizer` the
name of the regular expression does not matter. If does not define the
entity nor the intent, it is a human readable description for you
to remember what this regex is used for and is the title of the
corresponding pattern feature.

If you want to use the RegexFeaturizer you can also use the regex
features to improve the intent classification performance, for example,

> by defining a greet clause:

Reviewer 3: Do you think it's worth it to split this into 2 parts: regex for entity extraction and regex for intent classification?

Reviewer 3: Then we can easily add info about L308-314, as it only applies to entity extraction and not to intent classification (as all intent classifiers use the features)

Author: the features also apply to entities though. How about 
```
#### Regular Expression Features

#### Regular Expressions for Rule-Based Entity Extraction
```
Or is that too long a subtitle?

Author: Oh oh wait I see what you mean. Yes your version makes more sense.

```yaml
- regex: greet
  examples: |
  - hey[^\\s]*
```

Try to create your regular expressions in a way that they match as few
words as possible. E.g. using `hey[^\\s]*` instead of `hey.*`, as the
later one might match the whole message whereas the first one only
matches a single word.

When using the `RegexFeaturizer`, the regex features for entity extraction
are currently only supported by the `CRFEntityExtractor` and the
`DIETClassifier` component! Hence, other entity extractors, like
`MitieEntityExtractor` or `SpacyEntityExtractor` won't use the generated
features and their presence will not improve entity recognition for
these extractors. Currently, all intent classifiers make use of
available regex features.

> :::note
> 
> Regex features only define entities when used in combination with the `RegexEntityExtractor`. Otherwise they  
> don't define entities nor intents! They only provide patterns to
> help the classifier recognize entities and related intents. Hence, you
> still need to provide intent & entity examples as part of your
> training data!
> :::

Reviewer 3: ```suggestion
:::note

When using the `RegexFeaturizer`, a regular expression only provides a feature
that helps the model learn an association between intents/entities and inputs
that fit the regular expression. In order for the model to learn this association,
you must provide example inputs that include that regular expression! 
:::
```
Do you think this makes more sense? I'm trying to rephrase it as what to do when using features (versus when not using direct extraction)

### Lookup Tables

Lookup tables provide a convenient way to supply a list of entity
examples. The format is as follows:

```yaml
- lookup: <lookup table name>
  examples: |
  - <an entity>
  - <another entity>

> ```

Reviewer 3: similar to synonyms, let's use a real example

The name of the lookup table is subject to the same constraints as the
name of a regex feature.

> When lookup tables are supplied in training data, the contents are
> combined into a large regex pattern that looks for exact matches in the
> training examples. These regexes match over multiple tokens, so
> `lettuce wrap` would match `get me a lettuce wrap ASAP` as
> `[0 0 0 1 1 0]`. 

Reviewer 3: ```suggestion
When you supply a lookup table in your training data, the contents of that table
are combined into one large regular expression. This regex is used to check
each training example to see if it contains matches for entries in the
lookup table.
```

> These regexes are processed identically to the regular

Reviewer 3: ```suggestion
Lookup table regexes are processed identically to the
```

regex patterns directly specified in the training data and can be used
either with the [RegexFeaturizer](../nlu/components#regexfeaturizer)
or with the [RegexEntityExtractor](../nlu/components#regexentityextractor).

**note**

If you are using lookup tables in combination with the RegexFeaturizer, there must be a few examples of matches  
in your training data. Otherwise the model will not learn to use the
lookup table match features.

**warning**

You have to be careful when you add data to the lookup table.  
For example, if there are false positives or other noise in the table,
this can hurt performance. So make sure your lookup tables contain

> clean data.

Reviewer 3: These aren't an actual note + warning

If you are using lookup tables in combination with the RegexFeaturizer, there must be a few examples of matches  
in your training data. Otherwise the model will not learn to use the
lookup table match features.

**warning**

You have to be careful when you add data to the lookup table.  
For example, if there are false positives or other noise in the table,
this can hurt performance. So make sure your lookup tables contain
clean data.

> ## Stories

Reviewer: should we rather call this `Conversation training data (stories)`?

Author: Would stories & rules both fall under `Conversation training data`?

Reviewer: mhm that is a good point. Maybe it should be something along the lines of 

## Conversation training data
... You can do stories that use machine learning to identify patterns ... You can use rules to implement deterministic outcomes ... (prob needs better phrasing, but just for illustration)
### Stories
...
### Rules
...

Author: lmk if it makes sense now

Stories are composed of:
  
  - name: The name is arbitrary and not used in training; 
    you can use it as a human-readable reference for the story.
  - metadata: arbitrary and optional, not used in training, 
    you can use it to store relevant information about the story
    like e.g. the author

>   - a list of ``steps``: The user messages and actions that make up the story

Reviewer 3: name: here is confusing because it makes me think that is the key. maybe we do it by keys? 

```
Stories include the following keys: 

  - `story` (required): The story's name. The name is not used in training; 
     use it as a human-readable reference for the story.
  - `metadata` (optional): Extra information about the story. Not used in training; 
     you can use it to store relevant information about the story, e.g. the author.
  - a list of `steps` (required): The user messages and actions that make up the story.

Author: oh wow did not even see that I misnamed the name. Thanks

For example:

```yaml
stories:
- story: Greet the user
  metadata:
    author: Somebody
    key: value
  steps:
  # list of steps
  - intent: greet
  - action: utter_greet
```

Each step can be one of the following:

  - A [user message](#user-messages), represented by **intent** and **entities**.

>   - An [`or`](#or) statement, which includes two or more user messages under it

Reviewer 3: ```suggestion
  - An [or statement](#or), which includes two or more user messages under it
```
just to keep things consistent
